server/src/search.rs

```diff
@@ -634,3 +608,95 @@ fn highlight_matches(s: &String, search_tokens: &Vec<SearchToken>) -> String {
 
     s_highlighted.to_string()
 }
+
+// Parse the search against some known room formats and improve the
+// results display in this case. Room formats are hardcoded for now.
+fn parse_room_formats(
+    search_tokens: &Vec<SearchToken>,
+    hit: &MSHit,
+) -> Option<String> {
+    // Some building specific roomcode formats are determined by their building prefix
+    if search_tokens.len() == 2
+        && match search_tokens[0].s.as_str() {
+            "mi" => hit.id.starts_with("560") || hit.id.starts_with("561"),
+            "mw" => hit.id.starts_with("550") || hit.id.starts_with("551"),
+            "ph" => hit.id.starts_with("5101"),
+            "ch" => hit.id.starts_with("540"),
+            _ => false,
+        }
+        && !search_tokens[1].s.contains("@")
+        && hit.arch_name.is_some()
+        && hit
+            .arch_name
+            .as_ref()
+            .unwrap()
+            .starts_with(&search_tokens[1].s)
+    {
+        let arch_id = hit.arch_name.as_ref().unwrap().split("@").next().unwrap();
+        Some(format!(
+            "\u{0019}{} {}\u{0017}{}",
+            search_tokens[0].s.to_uppercase(),
+            arch_id.get(..search_tokens[1].s.len()).unwrap(),
+            arch_id.get(search_tokens[1].s.len()..).unwrap_or_default(),
+        ))
+
+    }
+    // If it doesn't match some precise room format, but the search is clearly
+    // matching the arch name and not the main name, then we highlight this arch name.
+    // This is intentionally still restrictive and considers only the first token,
+    // because we expect searches for arch names not to start with anything else.
+    else if (search_tokens.len() == 1
+             || (search_tokens.len() > 1 && search_tokens[0].s.len() >= 3))
+        //     Needs to be specific enough to be considered relevant ↑
+        && !hit.name.contains(&search_tokens[0].s) // No match in the name
+        && hit.parent_building.len() > 0 // Has parent information to show in query
+        && hit.arch_name.is_some()
+        && hit
+            .arch_name
+            .as_ref()
+            .unwrap()
+            .starts_with(&search_tokens[0].s)
+    {
+        // Exclude the part after the "@" if it's not in the query and use the
+        // building name instead, because this is probably more helpful
+        let (prefix, parsed_arch_id) = if search_tokens[0].s.contains("@") {
```

Comment:
honestly it is pretty hard to reason about this, because the else clause of this assignemnt is really large, but I dont know of a better alternative to simplify reasoning about this assignment, than to split this into its own function.

Reply:
Yes, it's a lot of ifs. I tried to comment some of them at least, but we should find a more structured approach in the future.